### PR TITLE
Cloud Architecture Design Guidance

### PR DESCRIPTION
**Pull request by Doommius**
# Pull Request Template

## What are you trying to address

- Write a detailed description of all changes - and if appropriate, why they are needed

This PR adds some best practices related to networking, DNS, and Ip allocations within azure. It relates to how to configure a saleable hub-and-spoke topology as well as a few points on how a hybrid deployment should be configured in terms of DNS

### CODE REVIEW

**Comment by Doommius** [comment]
@tompaana should be added as one of the Reviewers

File: docs/design/design-patterns/cloud-architecture-design-guidance/readme.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Cloud architecture Design Guidance
+
+
+In the below documet best practices about how network resources and architectures should be used in Azure Cloud enviorments.
+
+> **NOTE:** Always work with the relevant stakeholders to ensure that introducing new patterns provides the intended value.
+>
+> When working in an existing cloud environment, it is important to understand any current patterns and how they are used before making a change to them.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/cloud-architecture-design-guidance/readme.md:8]:
```suggestion
The following are some best practices when setting up and working with network resources in Azure Cloud environments.

> **NOTE:** When working in an existing cloud environment, it is important to understand any current patterns, and how they are used, before making a change to them. You should also work with the relevant stakeholders to make sure that any new patterns you introduce provide enough value to make the change.
```

File: docs/design/design-patterns/cloud-architecture-design-guidance/readme.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+* Scalability: The Hub-and-Spoke model can easily scale to accommodate additional spokes as the organization grows or as new applications and services are introduced.
+* Cost Savings: By centralizing shared services in the hub, organizations can reduce the costs associated with deploying and managing multiple instances of the same services across different VNets.
+
+Read more about [Hub-And-Spoke topology](https://learn.microsoft.com/en-us/azure/architecture/reference-architectures/hybrid-networking/hub-spoke?tabs=cli)
+
+It is recomended deploying hub/spoke in connection with [landing zones](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/landing-zone/) to ensure consistency accross all enviorments as well as guiderails to ensure a high level of security while giving devopers freedom within development enviorments.
+
+### Firewall and security
+
+When using a Hub-and-Spoke topology it is possible to deploy a centralized Firewall instance in the Hub that all outgoing traffic or traffic to/from certain vnets, this allows for centralized threat protection while minimizaing costs.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/cloud-architecture-design-guidance/readme.md:38]:
```suggestion
When using a Hub-and-Spoke topology it is possible to deploy a centralized Firewall instance in the Hub that all outgoing traffic or traffic to/from certain VNets, this allows for centralized threat protection while minimizing costs.
```

File: docs/design/design-patterns/cloud-architecture-design-guidance/readme.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+The best practice method for handling DNS in azure and in clouds in general means using the managed DNS services. There are quite a few benefits and resources designed to allow for secure and easy to implement and maintain configurations.
+
+* **DNS forwarding:** Set up DNS forwarding between your on-premises DNS servers and Azure DNS servers for name resolution across environments.
+* **Use Azure Private DNS zones for Azure resources:** Configure Azure Private DNS zones for your Azure resources to ensure name resolution is kept within the virtual network.
+
+Read more about [Hybrid/Multi-Cloud DNS infrasurecture](https://learn.microsoft.com/en-us/azure/architecture/hybrid/hybrid-dns-infra) and [Azure DNS infrazureture](https://learn.microsoft.com/en-us/azure/dns/)
+
+
+### IP allocation
+
```
Comment (TessFerrandez) [inline docs/design/design-patterns/cloud-architecture-design-guidance/readme.md:51]:
```suggestion
### IP allocation
```

File: docs/design/design-patterns/cloud-architecture-design-guidance/readme.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+When allocating IP address spaces to Azure Virtual Networks (VNets), it's essential to follow best practices for proper management, scalability, and to avoid potential issues. Here are some recommendations for IP allocation to VNets:
```
Comment (TessFerrandez) [inline docs/design/design-patterns/cloud-architecture-design-guidance/readme.md:53]:
```suggestion
When allocating IP address spaces to Azure Virtual Networks (VNets), it's essential to follow best practices for proper management, and scalability. 

Here are some recommendations for IP allocation to VNets:
```

File: docs/design/design-patterns/cloud-architecture-design-guidance/readme.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+* **Use subnets:** Divide your VNets into smaller subnets based on security, application, or environment requirements. This allows for better network management and security.
+* **Consider leaving a buffer between VNets:** While it's not strictly necessary, leaving a buffer between VNets can be beneficial in some cases, especially when you anticipate future growth or when you might need to merge VNets. This can help avoid re-addressing conflicts when expanding or merging networks.
+* **Reserve IP addresses:** Reserve a range of IP addresses within your VNet address space for critical resources or services. This ensures that they have a static IP address, which is essential for specific services or applications.
+* **Plan for hybrid scenarios:** If you're working in a hybrid environment with on-premises or multi-cloud networks, ensure that you plan for IP address allocation across all environments. This includes avoiding overlapping address spaces and reserving IP addresses for specific resources like VPN gateways or ExpressRoute circuits.
+
```
Comment (TessFerrandez) [inline docs/design/design-patterns/cloud-architecture-design-guidance/readme.md:64]:
```suggestion
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# Hybrid Resource Design Guidance
+
+
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:3]:
```suggestion
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+In the below documet best practices about how resources and architecture should be configured and used for Hybrid and Multi-Cloud enviorments.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:4]:
```suggestion
The following are best practices around how to design and configure resources, used for Hybrid and Multi-Cloud environments.
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+> **NOTE:** Always work with the relevant stakeholders to ensure that introducing new patterns provides the intended value.
+>
+> When working in an existing hybrid environment, it is important to understand any current patterns and how they are used before making a change to them.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:8]:
```suggestion
> **NOTE:** When working in an existing hybrid environment, it is important to understand any current patterns, and how they are used before making any changes.
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:9]:
```suggestion
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+## Hub and Spoke topology
+
+The Hub-and-Spoke topology doesn't change much when using cloud/hybrid if configured correctly, The main different is that the hub vnet is peering to the on-prem network via a Expressruote and that all traffic from Azure might exit via the expressroute and the on-prem internet connection.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:13]:
```suggestion
The Hub-and-Spoke topology doesn't change much when using cloud/hybrid if configured correctly, The main different is that the hub VNet is peering to the on-prem network via a ExpressRoute and that all traffic from Azure might exit via the ExpressRoute and the on-prem internet connection.
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+The generalized best practices are in  [cloud-architecture-design-guidance#Hub and Spoke topology](../cloud-architecture-design-guidance/readme.md#hub-and-spoke-topology)
+
+### IP allocation
+
+When working with Hybrid deployment extra care should be taken when planning IP allocation as there is a much greater risk of overlapping network ranges.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:19]:
```suggestion
When working with Hybrid deployment, take extra care when planning IP allocation as there is a much greater risk of overlapping network ranges.
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+The generalized best practices are in [cloud-architecture-design-guidance#ip-allocation](../cloud-architecture-design-guidance/readme.md#ip-allocation)
+
+Read more about this at [azure-best-practices/plan-for-ip-addressing](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:23]:
```suggestion
Read more about this in [Azure Best Practices Plan for IP Addressing](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```

File: docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+Choose the right ExpressRoute circuit: Select an appropriate SKU (Standard or Premium) and bandwidth based on your organization's requirements.
+Redundancy: Ensure redundancy by provisioning two ExpressRoute circuits in different peering locations.
+Monitoring: Use Azure Monitor and Network Performance Monitor (NPM) to monitor the health and performance of your ExpressRoute circuits.
+
+### DNS
+
+The generalized best practices are in [cloud-architecture-design-guidance#dns](../cloud-architecture-design-guidance/readme.md#dns)
+
+When using Azure DNS in a hybrid or multi-cloud enviorment it is important to ensure a consistent dns and forwarding configuration which ensures that records are automatically updated and that all DNS servers are aware of each other and know which server is the authoritative for the different records.
```
Comment (TessFerrandez) [inline docs/design/design-patterns/hybrid-architecture-design-guidance/readme.md:39]:
```suggestion
When using Azure DNS in a hybrid or multi-cloud environment it is important to ensure a consistent DNS and forwarding configuration which ensures that records are automatically updated and that all DNS servers are aware of each other and know which server is the authoritative for the different records.
```

**Review by TessFerrandez — COMMENTED** [review]
Thank you for the contribution.

Generally, may I recommend installing a spell checker in VS code and also a markdown lint extensions in VS code as this helps reduce common typos. I tried to catch them all but may have missed some.

I also think that the Cloud Architecture Design Guidance should perhaps be renamed to Network Architecture Guidance for Azure or something similar. Same with the hybrid, as this is not about the complete architecture but rather about the network specifics.

A third comment, and we don't have to change this now, but we could add this to the backlog. The design patterns sections seems to have unnecessary folders. Would be better, instead of having a folder with a readme, to just have the document straight in design patterns, and name it appropriately.

I will add a task to the backlog to make the structure more flat

**Comment by Doommius** [comment]
Noted, I'll change the naming in this pr to Network Architecture Guidance and I can create a 2nd pr after this is merged to update the folder structure to be flatter.

File: docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+Implementing a Hub-and-Spoke model in Azure offers several benefits:
+
+* Isolation and Segmentation: By dividing resources into separate spoke VNets, you can isolate and segment workloads, preventing any potential issues or security risks from affecting other parts of the network.
+* Centralized Management: The hub VNet acts as a single point of management for shared services, making it easier to maintain, monitor, and enforce policies across the network.
+* Simplified Connectivity: VNet peering enables seamless communication between the hub and spoke VNets without the need for complex routing or additional gateways, reducing latency and management overhead.
+* Scalability: The Hub-and-Spoke model can easily scale to accommodate additional spokes as the organization grows or as new applications and services are introduced.
+* Cost Savings: By centralizing shared services in the hub, organizations can reduce the costs associated with deploying and managing multiple instances of the same services across different VNets.
+
+Read more about [Hub-And-Spoke topology](https://learn.microsoft.com/en-us/azure/architecture/reference-architectures/hybrid-networking/hub-spoke?tabs=cli)
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md:28]:
```suggestion
Read more about [Hub-And-Spoke topology](https://learn.microsoft.com/azure/architecture/reference-architectures/hybrid-networking/hub-spoke?tabs=cli)
```

File: docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+When deploying hub/spoke, it is recommended that you do so in connection with [landing zones](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/landing-zone/). This ensures consistency across all environments as well as guardrails to ensure a high level of security while giving developers freedom within development environments.
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md:30]:
```suggestion
When deploying hub/spoke, it is recommended that you do so in connection with [landing zones](https://learn.microsoft.com/azure/cloud-adoption-framework/ready/landing-zone/). This ensures consistency across all environments as well as guardrails to ensure a high level of security while giving developers freedom within development environments.
```

File: docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+### Firewall and security
+
+When using a Hub-and-Spoke topology it is possible to deploy a centralized Firewall instance in the Hub that all outgoing traffic or traffic to/from certain VNets, this allows for centralized threat protection while minimizing costs.
+
+### DNS
+
+The best practices for handling DNS in azure and in clouds in general include using the managed DNS services. Some of the benefits of using the managed DNS services is that the resources are designed to be secure, and easy to deploy and configure.
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md:38]:
```suggestion
The best practices for handling DNS in Azure, and in cloud environments in general, include using managed DNS services. Some of the benefits of using managed DNS services is that the resources are designed to be secure, easy to deploy and configure.
```

File: docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+* **DNS forwarding:** Set up DNS forwarding between your on-premises DNS servers and Azure DNS servers for name resolution across environments.
+* **Use Azure Private DNS zones for Azure resources:** Configure Azure Private DNS zones for your Azure resources to ensure name resolution is kept within the virtual network.
+
+Read more about [Hybrid/Multi-Cloud DNS infrastructure](https://learn.microsoft.com/en-us/azure/architecture/hybrid/hybrid-dns-infra) and [Azure DNS infratructure](https://learn.microsoft.com/en-us/azure/dns/)
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md:43]:
```suggestion
Read more about [Hybrid/Multi-Cloud DNS infrastructure](https://learn.microsoft.com/azure/architecture/hybrid/hybrid-dns-infra) and [Azure DNS infratructure](https://learn.microsoft.com/azure/dns/)
```

File: docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+* **Consider leaving a buffer between VNets:** While it's not strictly necessary, leaving a buffer between VNets can be beneficial in some cases, especially when you anticipate future growth or when you might need to merge VNets. This can help avoid re-addressing conflicts when expanding or merging networks.
+* **Reserve IP addresses:** Reserve a range of IP addresses within your VNet address space for critical resources or services. This ensures that they have a static IP address, which is essential for specific services or applications.
+* **Plan for hybrid scenarios:** If you're working in a hybrid environment with on-premises or multi-cloud networks, ensure that you plan for IP address allocation across all environments. This includes avoiding overlapping address spaces and reserving IP addresses for specific resources like VPN gateways or ExpressRoute circuits.
+
+Read more at [azure-best-practices/plan-for-ip-addressing](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-azure/readme.md:61]:
```suggestion
Read more at [azure-best-practices/plan-for-ip-addressing](https://learn.microsoft.com/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```

File: docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+The generalized best practices are in  [Network Architecture Guidance for Azure#Hub and Spoke topology](../network-architecture-guidance-for-azure/readme.md#hub-and-spoke-topology)
+
+### IP allocation
+
+When working with Hybrid deployment, take extra care when planning IP allocation as there is a much greater risk of overlapping network ranges.
+
+The general best practices are available in the [Network Architecture Guidance for Azure#ip-allocation](../network-architecture-guidance-for-azure/readme.md#ip-allocation)
+
+Read more about this in [Azure Best Practices Plan for IP Addressing](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md:19]:
```suggestion
Read more about this in [Azure Best Practices Plan for IP Addressing](https://learn.microsoft.com/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-ip-addressing)
```

File: docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+### ExpressRoute
+
+Environments using Express often tunnel all traffic from Azure via a private link (ExpressRoute) to an on-prem location. This imposes a few problems when working with PAAS offerings as not all of them connect via their respective private endpoint and instead use an external IP for outgoing connections, or some PAAS to PASS traffic occur internally in azure and won't function with disabled public networks.
+
+Two notable services here are data planes copies of Storage accounts and a lot of the services not supporting private-endpoints.
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md:25]:
```suggestion
Two notable services here are data planes copies of storage accounts and a lot of the services not supporting private endpoints.
```

File: docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+Redundancy: Ensure redundancy by provisioning two ExpressRoute circuits in different peering locations.
+Monitoring: Use Azure Monitor and Network Performance Monitor (NPM) to monitor the health and performance of your ExpressRoute circuits.
+
+### DNS
+
+General best practices are available in [Network Architecture Guidance for Azure#dns](../network-architecture-guidance-for-azure/readme.md#dns)
+
+When using Azure DNS in a hybrid or multi-cloud environment it is important to ensure a consistent DNS and forwarding configuration which ensures that records are automatically updated and that all DNS servers are aware of each other and know which server is the authoritative for the different records.
+
+Read more about [Hybrid/Multi-Cloud DNS infrastructure](https://learn.microsoft.com/en-us/azure/architecture/hybrid/hybrid-dns-infra)
```
Comment (tompaana) [inline docs/design/design-patterns/network-architecture-guidance-for-hybrid/readme.md:37]:
```suggestion
Read more about [Hybrid/Multi-Cloud DNS infrastructure](https://learn.microsoft.com/azure/architecture/hybrid/hybrid-dns-infra)
```

**Review by tompaana — APPROVED** [review]
Approved w/ suggestions.

**Comment by tompaana** [comment]
> Thank you for the contribution.
> 
> Generally, may I recommend installing a spell checker in VS code and also a markdown lint extensions in VS code as this helps reduce common typos. I tried to catch them all but may have missed some.
> 
> I also think that the Cloud Architecture Design Guidance should perhaps be renamed to Network Architecture Guidance for Azure or something similar. Same with the hybrid, as this is not about the complete architecture but rather about the network specifics.
> 
> A third comment, and we don't have to change this now, but we could add this to the backlog. The design patterns sections seems to have unnecessary folders. Would be better, instead of having a folder with a readme, to just have the document straight in design patterns, and name it appropriately.
> 
> I will add a task to the backlog to make the structure more flat

What @TessFerrandez said. Although the text looks quite polished, I'd suggest revisiting the capitalisation; some words that start with a capital letter without justification. My go to source to check is our official documentation, but that too can have varying ways of formatting and spelling things unfortunately.

**Comment by Doommius** [comment]
@tompaana

I've tried my best to follow the same formatting as the rest of the docs.

Headers and List headers use capitalize the first letter of each word except conjunctions. However, looking at the [style guide](https://learn.microsoft.com/en-us/style-guide/capitalization) this is the incorrect practice. Should I revise it to follow the style guide or keep it matching to the rest of the playbook?

**Comment by tompaana** [comment]
> @tompaana
> 
> I've tried my best to follow the same formatting as the rest of the docs.
> 
> Headers and List headers use capitalize the first letter of each word except conjunctions. However, looking at the [style guide](https://learn.microsoft.com/en-us/style-guide/capitalization) this is the incorrect practice. Should I revise it to follow the style guide or keep it matching to the rest of the playbook?

Like I said, it does look quite clean. In general, and in my opinion, consistency is key when it comes to headers. I was more interesting about the paragraphs e.g.,
image

This could be right, could be not. Also I saw "Firewall" in a middle of sentence when the word was used as a common noun (vs. "Azure Firewall").

**Review by TessFerrandez — APPROVED** [review]
Thank you much for the contribution, very useful stuff

**Pull request merged**